File: MovieApp/app/src/main/java/com/anhhoang/popularmovies/data/FavoriteMovieContentProvider.java
package com.anhhoang.popularmovies.data;

import android.content.ContentProvider;
import android.content.ContentUris;
import android.content.ContentValues;
import android.content.Context;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import static com.anhhoang.popularmovies.data.FavoriteMovieContract.FavoriteMovieEntry.TABLE_NAME;

/*
 * Copyright (C) 2013 The Android Open Source Project
 */

/**
 * Created by AnhHo on 3/23/2017.
 */

public class FavoriteMovieContentProvider extends ContentProvider {
    public static final int FAVORITE_MOVIES = 100;
    public static final int FAVORITE_MOVIES_WITH_ID = 101;

    private static final UriMatcher sUriMatcher = buildUriMatcher();
    private FavoriteMovieDbHelper mFavoriteMovieDbHelper;

    public static UriMatcher buildUriMatcher() {
        UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);

        uriMatcher.addURI(FavoriteMovieContract.AUTHORITY, FavoriteMovieContract.PATH_FAV_MOVIES, FAVORITE_MOVIES);
        uriMatcher.addURI(FavoriteMovieContract.AUTHORITY, FavoriteMovieContract.PATH_FAV_MOVIES + "/#", FAVORITE_MOVIES_WITH_ID);

        return uriMatcher;
    }

    @Override
    public boolean onCreate() {
        // Not creating global Realm Db access, because we want to close realm instance each time in case of memory leaking
        Context context = getContext();
        mFavoriteMovieDbHelper = new FavoriteMovieDbHelper(context);
        return true;
    }

    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        // Due to limitation of Realm.io we won't be able to use Projection, selection and etc.
        int match = sUriMatcher.match(uri);
        SQLiteDatabase db = mFavoriteMovieDbHelper.getReadableDatabase();
        Cursor cursor;
        switch (match) {
            case FAVORITE_MOVIES:
                cursor = db.query(
                        TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder
                );
                break;
            case FAVORITE_MOVIES_WITH_ID:
                long id = Long.parseLong(uri.getLastPathSegment());
                cursor = db.query(
                        TABLE_NAME,
                        projection,
                        FavoriteMovieContract.FavoriteMovieEntry._ID + "=?",
                        new String[]{String.valueOf(id)},
                        null,
                        null,
                        null
                );
                break;
            default:
                throw new UnsupportedOperationException("Not supported uri: " + uri);
        }

        cursor.setNotificationUri(getContext().getContentResolver(), uri);
        return cursor;
    }

    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        int match = sUriMatcher.match(uri);
        Uri uriResult = null;
        final SQLiteDatabase database = mFavoriteMovieDbHelper.getWritableDatabase();
        switch (match) {
            case FAVORITE_MOVIES:
                long id = database.insert(TABLE_NAME, null, values);
                if (id > 0) {
                    uriResult = ContentUris.withAppendedId(uri, id);
                } else {
                    throw new SQLException("Failed to insert row into " + uri);
                }
                break;
            default:
                throw new UnsupportedOperationException("Not supported uri: " + uri);
        }
        getContext().getContentResolver().notifyChange(uri, null);

        return uriResult;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        int match = sUriMatcher.match(uri);
        int result = 0;
        SQLiteDatabase db = mFavoriteMovieDbHelper.getWritableDatabase();
        switch (match) {
            case FAVORITE_MOVIES_WITH_ID:
                String id = uri.getLastPathSegment();

                result = db.delete(TABLE_NAME, FavoriteMovieContract.FavoriteMovieEntry._ID + "=?", new String[]{id});
                break;
            default:
                throw new UnsupportedOperationException("Not supported uri: " + uri);
        }
        if (result != 0) {
            getContext().getContentResolver().notifyChange(uri, null);
        }

        return result;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        throw new UnsupportedOperationException("Not supported operation");
    }
}

File: MovieApp/app/src/main/java/com/anhhoang/popularmovies/data/FavoriteMovieContentProvider.kt
package com.anhhoang.popularmovies.data

import android.content.ContentProvider
import android.content.ContentUris
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.net.Uri
import android.provider.BaseColumns._ID

import com.anhhoang.popularmovies.data.FavoriteMovieContract.FavoriteMovieEntry.Companion.TABLE_NAME

/*
 * Copyright (C) 2013 The Android Open Source Project
 */

/**
 * Created by AnhHo on 3/23/2017.
 */

class FavoriteMovieContentProvider : ContentProvider() {
    private lateinit var mFavoriteMovieDbHelper: FavoriteMovieDbHelper

    override fun onCreate(): Boolean {
        // Not creating global Realm Db access, because we want to close realm instance each time in case of memory leaking
        mFavoriteMovieDbHelper = FavoriteMovieDbHelper(context)
        return true
    }

    override fun query(uri: Uri, projection: Array<String>?, selection: String?, selectionArgs: Array<String>?, sortOrder: String?): Cursor? {
        // Due to limitation of Realm.io we won't be able to use Projection, selection and etc.
        val match = sUriMatcher.match(uri)
        val db = mFavoriteMovieDbHelper.readableDatabase
        val cursor: Cursor
        when (match) {
            FAVORITE_MOVIES -> cursor = db.query(
                    TABLE_NAME,
                    projection,
                    selection,
                    selectionArgs, null, null,
                    sortOrder
            )
            FAVORITE_MOVIES_WITH_ID -> {
                val id = java.lang.Long.parseLong(uri.lastPathSegment)
                cursor = db.query(
                        TABLE_NAME,
                        projection,
                        _ID + "=?",
                        arrayOf(id.toString()), null, null, null
                )
            }
            else -> throw UnsupportedOperationException("Not supported uri: " + uri)
        }

        cursor.setNotificationUri(context.contentResolver, uri)
        return cursor
    }

    override fun getType(uri: Uri): String? {
        return null
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val match = sUriMatcher.match(uri)
        val uriResult: Uri?
        val database = mFavoriteMovieDbHelper.writableDatabase
        when (match) {
            FAVORITE_MOVIES -> {
                val id = database.insert(TABLE_NAME, null, values)
                if (id > 0) {
                    uriResult = ContentUris.withAppendedId(uri, id)
                } else {
                    throw SQLException("Failed to insert row into " + uri)
                }
            }
            else -> throw UnsupportedOperationException("Not supported uri: " + uri)
        }
        context.contentResolver.notifyChange(uri, null)

        return uriResult
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<String>?): Int {
        val match = sUriMatcher.match(uri)
        val result: Int
        val db = mFavoriteMovieDbHelper.writableDatabase
        when (match) {
            FAVORITE_MOVIES_WITH_ID -> {
                val id = uri.lastPathSegment

                result = db.delete(TABLE_NAME, _ID + "=?", arrayOf(id))
            }
            else -> throw UnsupportedOperationException("Not supported uri: " + uri)
        }
        if (result != 0) {
            context.contentResolver.notifyChange(uri, null)
        }

        return result
    }

    override fun update(uri: Uri, values: ContentValues?, selection: String?, selectionArgs: Array<String>?): Int {
        throw UnsupportedOperationException("Not supported operation")
    }

    companion object {
        val FAVORITE_MOVIES = 100
        val FAVORITE_MOVIES_WITH_ID = 101

        private val sUriMatcher = buildUriMatcher()

        fun buildUriMatcher(): UriMatcher {
            val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)

            uriMatcher.addURI(FavoriteMovieContract.AUTHORITY, FavoriteMovieContract.PATH_FAV_MOVIES, FAVORITE_MOVIES)
            uriMatcher.addURI(FavoriteMovieContract.AUTHORITY, FavoriteMovieContract.PATH_FAV_MOVIES + "/#", FAVORITE_MOVIES_WITH_ID)

            return uriMatcher
        }
    }
}
